### PR TITLE
Fix scheduler race condition.

### PR DESCRIPTION
**Pull request by davinchia**
## What
Implements concurrent hash set fix from #4378 .

I left code comments throughout, so should be relatively easy to follow.

## Recommended reading order
1. `JobSubmitter.java`
2. `JobSubmitterTest.java`

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [ ] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] `README.md`
    - [ ] `docs/SUMMARY.md` if it's a new connector
    - [ ] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [ ] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md` contains a reference to the new connector
    - [ ] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

<details><summary> <strong> Connector Generator checklist </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-scheduler/app/src/test/java/io/airbyte/scheduler/app/JobSubmitterTest.java [hunk 1]
```diff
@@ -226,4 +230,68 @@ void testMDC() throws Exception {
...
+        });
+      }
+
+      simulatedJobSubmitterPool.shutdownNow();
+      verify(persistence, Mockito.times(submitCounter.get())).getNextJob();
+      verify(jobSubmitter, Mockito.times(1)).submitJob(Mockito.any());
+    }
+
+    @Test
+    public void testFailureShouldUnlockId() throws Exception {
```
Comment (cgardens) [inline airbyte-scheduler/app/src/test/java/io/airbyte/scheduler/app/JobSubmitterTest.java:264]: i think the name of this test and the test on line  280 are switched? this one tests the success case and the other tests the failure case.

File: airbyte-scheduler/app/src/test/java/io/airbyte/scheduler/app/JobSubmitterTest.java [hunk 2]
```diff
@@ -226,4 +230,68 @@ void testMDC() throws Exception {
     assertTrue(MDC.getCopyOfContextMap().isEmpty());
   }
 
+  @Nested
+  class OnlyOneJobIdRunning {
+
+    @Test
+    public void testMultipleSubmissionShouldFail() throws Exception {
```
Comment (cgardens) [inline airbyte-scheduler/app/src/test/java/io/airbyte/scheduler/app/JobSubmitterTest.java:237]: i think the name of this one is a bit misleading or maybe add a comment explaining what you're testing. this was a bit hard for me to parse. if i'm understanding it correctly it is testing that while a job is running or queued, the submitter still runs but if it tries to submit the same job again subsequent submissions don't happen. thus `getNextJob` happens more than once but `submitJob` should happen exactly once.

**Review by cgardens — APPROVED** [review]

**Review by jrhizor — APPROVED** [review]

File: airbyte-scheduler/app/src/test/java/io/airbyte/scheduler/app/JobSubmitterTest.java (hunk 1 above)
Comment (davinchia) [reply airbyte-scheduler/app/src/test/java/io/airbyte/scheduler/app/JobSubmitterTest.java:264]: Ugh. Thanks!

File: airbyte-scheduler/app/src/test/java/io/airbyte/scheduler/app/JobSubmitterTest.java (hunk 2 above)
Comment (davinchia) [reply airbyte-scheduler/app/src/test/java/io/airbyte/scheduler/app/JobSubmitterTest.java:237]: Exactly that. I will add comments.

**Pull request merged**